compiler/src/test/java/com/squareup/anvil/compiler/codegen/CodeGenerationExtensionTest.kt

```diff
@@ -83,36 +81,155 @@ class CodeGenerationExtensionTest {
+
+    val codeGenerator = simpleCodeGenerator(NO_SOURCE_TRACKING) { clazz ->
+
+      if (clazz.shortName == "Abc") {
+        return@simpleCodeGenerator null
       }
 
-      override fun generateCode(
-        codeGenDir: File,
-        module: ModuleDescriptor,
-        projectFiles: Collection<KtFile>,
-      ): Collection<GeneratedFile> {
-        generateCodeCalled = true
-        return emptyList()
+      """
+        package com.squareup.test
+          
+        class Abc
+      """.trimIndent()
+    }
+
+    //language=kotlin
+    val componentInterface = """
+      package com.squareup.test
+
+      interface ComponentInterface
+    """.trimIndent()
+
+    compile(
+      componentInterface,
+      codeGenerators = listOf(codeGenerator),
+      trackSourceFiles = true,
+    ) {
+      assertThat(codeGenerator.isApplicableCalls).isEqualTo(1)
+      assertThat(codeGenerator.getCallsForInput(componentInterface)).isEqualTo(1)
+
+      assertThat(exitCode).isEqualTo(INTERNAL_ERROR)
+
+      val abcPath = outputDirectory
+        .resolveSibling("build/anvil/com/squareup/test/Abc.kt")
+        .absolutePath
+
+      assertThat(messages).contains(
+        """
+        |Source file tracking is enabled but this generated file is not tracking them.
+        |Please report this issue to the code generator's maintainers.
+        |
+        |The file was generated by ${SimpleCodeGenerator::class}.
+        |The file is: $abcPath
+        |
+        |To stop this error, enable the `trackSourceFiles` property in the Anvil Gradle extension:
+        |
+        |   // build.gradle(.kts)
+        |   anvil {
+        |     trackSourceFiles = true
+        |   }
+        |
+        |or enable the property in `gradle.properties`:
+        |
+        |   # gradle.properties
+        |   com.squareup.anvil.trackSourceFiles=true
+        |
+        """.trimMargin(),
+
+      )
+    }
+  }
+
+  @Test fun `a code generator that tracks an empty source list can use trackSourceFiles`() {
+
+    val codeGenerator = simpleCodeGenerator(TRACKING_WITH_NO_SOURCES) { clazz ->
+
+      if (clazz.shortName == "Abc") {
+        return@simpleCodeGenerator null
       }
+
+      """
+        package com.squareup.test
+          
+        class Abc
+      """.trimIndent()
     }
 
+    //language=kotlin
+    val componentInterface = """
+      package com.squareup.test
+
+      interface ComponentInterface
+    """.trimIndent()
+
     compile(
+      componentInterface,
+      codeGenerators = listOf(codeGenerator),
+      trackSourceFiles = true,
+    ) {
+      assertThat(codeGenerator.isApplicableCalls).isEqualTo(1)
+      assertThat(codeGenerator.getCallsForInput(componentInterface)).isEqualTo(1)
+
+      assertThat(exitCode).isEqualTo(OK)
+      assertThat(classLoader.loadClass("com.squareup.test.Abc")).isNotNull()
+    }
+  }
+
+  @Test
+  fun `a code generator that does not track sources can compile with trackSourceFiles disabled`() {
```

Comment:
could you also add a test for the case where trackSourceFIles is enabled and there's at least one source file tracked?

Reply:
That's the current base case, so it's tested over 1000 times, but I see the logic of it. 😄 

Added